packages/auth/lib/Users.js

```diff
@@ -40,9 +41,91 @@ const {
   FRONTEND_BASE_URL
 } = process.env
 
-const signIn = async (_email, context, pgdb, req, consents) => {
+// in order of preference
+const enabledFirstFactors = async (email, pgdb) => {
+  const userWithDevices = await pgdb.query(`
+    SELECT
+      u.*,
+      jsonb_agg(d.*) as devices
+    FROM
+      users u
+    LEFT JOIN
+      devices d
+      ON u.id = d."userId"
+    WHERE
+      u.email = :email
+    GROUP BY
+      u.id
+  `, {
+    email
+  })
+    .then(result => {
+      const user = result[0]
+      if (!user) {
+        return user
+      }
+      return {
+        ...user,
+        devices: user.devices
+          .filter(Boolean)
+      }
+    })
+
+  const { APP, EMAIL_TOKEN } = TokenTypes
+
+  const enabledTokenTypes = [EMAIL_TOKEN]
+  if (userWithDevices && userWithDevices.devices.length > 0) {
+    enabledTokenTypes.push(APP)
+  }
+
+  const { preferredFirstFactor } = userWithDevices || {}
+  return enabledTokenTypes
+    .sort((a, b) => {
+      if (a === b) {
+        return 0
+      }
+      // preferred always wins
+      if (a === preferredFirstFactor) {
+        return -1
+      }
+      if (b === preferredFirstFactor) {
+        return 1
+      }
+      // others are always before EMAIL_TOKEN
+      if (b === EMAIL_TOKEN) {
+        return -1
+      }
+      if (a === EMAIL_TOKEN) {
+        return 1
+      }
+      // undefined order for others
+      return 0
+    })
+}
+
+const setPreferredFirstFactor = async (user, tokenType = null, pgdb) => {
+  const enabledTokenTypes = await enabledFirstFactors(user._raw.email, pgdb)
+  if (tokenType && enabledTokenTypes.indexOf(tokenType) === -1) {
+    throw new TokenTypeNotEnabledError({ tokenType })
```

Comment:
Always allow to set preference to known token types ('APP', 'EMAIL'). FE will relax UI restrictions, letting users set the app preference independently of whether a device is available at the exact moment.